internal/pkg/composable/providers/kubernetesleaderelection/kubernetes_leaderelection_test.go

```diff
@@ -0,0 +1,229 @@
+// 4. We will force the lease to be acquired by the agent1 again. To avoid the agent2 reacquiring it multiple times,
+// we will stop this provider and make sure the agent1 can reacquire it.
+func TestNewLeaderElectionManager(t *testing.T) {
+	client := k8sfake.NewSimpleClientset()
+
+	lease := createLease()
+	// create the lease that leader election will be using
+	err := applyLease(client, lease, true)
+	require.NoError(t, err)
+
+	// Create the provider
+	logger := logp.NewLogger("test_leaderelection")
+
+	leaseDuration := 3
+	leaseRenewDeadline := 2
+	leaseRetryPeriod := 1
+
+	c := map[string]interface{}{
+		"leader_lease":         leaseName,
+		"leader_leaseduration": leaseDuration,
+		"leader_renewdeadline": leaseRenewDeadline,
+		"leader_retryperiod":   leaseRetryPeriod,
+	}
+	cfg, err := config.NewConfigFrom(c)
+	require.NoError(t, err)
+
+	getK8sClientFunc = func(kubeconfig string, opt autodiscoverK8s.KubeClientOptions) (kubernetes.Interface, error) {
+		return client, nil
+	}
+	require.NoError(t, err)
+
+	podNames := [2]string{"agent1", "agent2"}
+	cancelFuncs := [2]context.CancelFunc{}
+
+	done := make(chan int, 1)
+
+	// Create two leader election providers representing two agents running
+	for i := 0; i < 2; i++ {
+		p, err := ContextProviderBuilder(logger, cfg, true)
+		require.NoError(t, err)
+
+		ctx, cancel := context.WithCancel(context.Background())
+		cancelFuncs[i] = cancel
+		defer cancel()
+
+		comm := ctesting.NewContextComm(ctx)
+
+		err = os.Setenv("POD_NAME", podNames[i])
+		if err != nil {
+			require.FailNow(t, "Failed to set pod name environment variable.")
+		}
+		go func() {
+			_ = p.Run(ctx, comm)
+		}()
+
+		if i == 1 {
+			break
+		}
+
+		// We need to wait for the first agent to acquire the lease, so we can POD_NAME environment variable again
+		go func() {
+			expectedLeader := leaderElectorPrefix + podNames[i]
+			for {
+				holder, err := getLeaseHolder(client)
+				require.NoError(t, err)
+
+				if holder == expectedLeader {
+					done <- 1
+					break
+				}
+			}
+		}()
+
+		select {
+		case <-done:
+		case <-time.After(time.Duration(leaseDuration+leaseRetryPeriod) * 30 * time.Second):
+			require.FailNow(t, "Timeout"+
+				" while waiting for the first pod to acquire the lease. This should not happen. Consider increasing "+
+				"the timeout.")
+		}
+	}
+
+	go func() {
+		// At this point the current holder is agent1. Let's change it to agent2.
+		for {
+			// Force the lease to be applied again, so a new leader is elected.
+			intermediateHolder := "does-not-matter"
+			lease.Spec.HolderIdentity = &intermediateHolder
+			err = applyLease(client, lease, false)
+			require.NoError(t, err)
+
+			var currentHolder string
+			for {
+				currentHolder, err = getLeaseHolder(client)
+				require.NoError(t, err)
+
+				// In this case, we already have an agent as holder
+				if currentHolder == leaderElectorPrefix+podNames[0] || currentHolder == leaderElectorPrefix+podNames[1] {
```

Comment:
I am not adding a timeout to this for cycle because this is a sure thing to happen, since there are only 2 pods running to acquire the lease.